server/src/test/scala/org/http4s/server/middleware/authentication/AuthenticationSpec.scala

```diff
@@ -137,93 +154,107 @@ class AuthenticationSpec extends Http4sSpec {
-      val params: NonEmptyList[(String, String)] = NonEmptyList(
-        "username" -> username, "realm" -> realm, "nonce" -> nonce,
-        "uri" -> uri, "qop" -> qop, "nc" -> nc, "cnonce" -> cnonce,
-        "response" -> response, "method" -> method
+      val response =
+        DigestUtil.computeResponse(method, username, realm, password, uri, nonce, nc, cnonce, qop)
+      val params: NonEmptyList[(String, String)] = NonEmptyList.of(
+        "username" -> username,
+        "realm" -> realm,
+        "nonce" -> nonce,
+        "uri" -> uri,
+        "qop" -> qop,
+        "nc" -> nc,
+        "cnonce" -> cnonce,
+        "response" -> response,
+        "method" -> method
       )
       val header = Authorization(Credentials.AuthParams("Digest".ci, params))
 
       val req2 = Request[IO](uri = Uri(path = "/"), headers = Headers(header))
-      val res2 = digest.orNotFound(req2).run
+      val res2 = digest.orNotFound(req2).unsafeRunSync
 
       if (withReplay) {
-        val res3 = digest.orNotFound(req2).run
+        val res3 = digest.orNotFound(req2).unsafeRunSync
         (res2, res3)
       } else
         (res2, null)
     }
 
     "Respond to a request with correct credentials" in {
-      val digestAuthService = DigestAuth(realm, authStore)(service)
+      val digestAuthService = digestAuthMiddleware(service)
       val challenge = doDigestAuth1(digestAuthService)
 
       (challenge match {
-        case Challenge("Digest", realm, _) => true
+        case Challenge("Digest", `realm`, _) => true
         case _ => false
-      }) must_== true
+      }) must beTrue
 
-      val (res2, res3) = doDigestAuth2(digestAuthService, challenge, true)
+      val (res2, res3) = doDigestAuth2(digestAuthService, challenge, withReplay = true)
 
-      res2.status must_== (Ok)
+      res2.status must_=== Ok
 
       // Digest prevents replay
-      res3.status must_== (Unauthorized)
+      res3.status must_=== Unauthorized
 
       ok
     }
 
     "Respond to many concurrent requests while cleaning up nonces" in {
       val n = 100
-      val sched = Executors.newFixedThreadPool(4)
-      val digestAuthService = DigestAuth(realm, authStore, 2.millis, 2.millis)(service)
-      val tasks = (1 to n).map(i =>
-        Task {
-          val challenge = doDigestAuth1(digestAuthService)
-          (challenge match {
-            case Challenge("Digest", realm, _) => true
-            case _ => false
-          }) must_== true
-          val res = doDigestAuth2(digestAuthService, challenge, false)._1
-          // We don't check whether res.status is Ok since it may not
-          // be due to the low nonce stale timer.  Instead, we check
-          // that it's found.
-          res.status mustNotEqual (NotFound)
-        }(sched))
-      Task.gatherUnordered(tasks).run
+      implicit val ec = ExecutionContext.fromExecutor(Executors.newFixedThreadPool(4))
```

Comment:
Your threads are showing. This isn't you, but should get fixed.

Reply:
Wow, am I glad to find these.